samples/RefreshFlow/AngularApp/src/core/auth.service.ts

```diff
@@ -0,0 +1,164 @@
+import { ProfileModel } from './models/profile-model';
+import { AuthStateModel } from './models/auth-state-model';
+import { AuthTokenModel } from './models/auth-tokens-model';
+import { RegisterModel } from './models/register-model';
+import { LoginModel } from './models/login-model';
+
+const jwtDecode = require('jwt-decode');
+
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/mergeMap';
+import 'rxjs/add/operator/first';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/filter';
+
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/interval';
+import 'rxjs/add/observable/throw';
+
+@Injectable()
+export class AuthService {
+
+    private initalState: AuthStateModel = { profile: null, tokens: null, authReady: false };
+    private authReady$ = new BehaviorSubject<boolean>(false);
+    state: BehaviorSubject<AuthStateModel>;
+    private refreshSubscription$: Subscription;
+
+    state$: Observable<AuthStateModel>;
+    tokens$: Observable<AuthTokenModel>;
+    profile$: Observable<ProfileModel>;
+    loggedIn$: Observable<boolean>;
+
+    constructor(
+        private http: Http,
+    ) {
+        this.state = new BehaviorSubject<AuthStateModel>(this.initalState);
+        this.state$ = this.state.asObservable();
+    }
+
+    init() {
+        this.tokens$ = this.state.filter(state => state.authReady)
+            .map(state => state.tokens);
+
+        this.profile$ = this.state.filter(state => state.authReady)
+            .map(state => state.profile);
+
+        this.loggedIn$ = this.tokens$.map(tokens => !!tokens);
+
+        return this.startupTokenRefresh()
+            .do(() => this.scheduleRefresh());
+    }
+
+    register(data: RegisterModel): Observable<Response> {
+        return this.http.post('http://localhost:5056/account/register', data)
+            .catch(res => Observable.throw(res.json()));
+    }
+
+    login(user: LoginModel) {
+        return this.getTokens(user, 'password')
+            .catch(res => Observable.throw(res.json()))
+            .do(res => this.scheduleRefresh());
+    }
+
+    logout() {
+        this.updateState({ profile: null, tokens: null });
+        if (this.refreshSubscription$) {
+            this.refreshSubscription$.unsubscribe();
+        }
+        this.removeToken();
+    }
+
+    isInRole(usersRole: string): Observable<boolean> {
+        return this.profile$
+            .map(profile => profile.role.find(role => role === usersRole) !== undefined);
+    }
+
+    refreshTokens() {
+        return this.state.first()
+            .map(state => state.tokens)
+            .flatMap(tokens => this.getTokens({ refresh_token: tokens.refresh_token }, 'refresh_token')
+                .catch(error => Observable.throw('Session Expired'))
+            );
+    }
+
+    private storeToken(tokens: AuthTokenModel) {
+        localStorage.setItem('auth-tokens', JSON.stringify(tokens));
+    }
+    private retrieveTokens() {
```

Comment:
This is an example of inconsistent vertical whitespace. I would expect a consistent use of one space between functions.